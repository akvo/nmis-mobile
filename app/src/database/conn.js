import { Platform } from 'react-native';
import { Asset } from 'expo-asset';
import * as FileSystem from 'expo-file-system';
import * as SQLite from 'expo-sqlite';

const openDatabase = () => {
  if (Platform.OS === 'web') {
    return {
      transaction: () => {
        return {
          executeSql: () => {},
        };
      },
    };
  }
  const db = SQLite.openDatabase('db.db');
  return db;
};

const init = openDatabase();

const tx = (db, query, params = []) => {
  return new Promise((resolve, reject) => {
    db.transaction(
      (transaction) => {
        if (Array.isArray(query)) {
          const results = [];
          query.forEach(async (q) => {
            transaction.executeSql(q, params, (_, resultSet) => {
              results.push(resultSet); // Store the result set in the array
            });
          });
          resolve(results);
        } else {
          transaction.executeSql(
            query,
            params,
            (_, resultSet) => {
              resolve(resultSet);
            },
            (_, error) => {
              reject(error);
              return false; // Rollback the transaction
            },
          );
        }
      },
      (error) => {
        reject(error);
      },
    );
  });
};

const openDBfile = async (databaseFile, databaseName) => {
  if (!(await FileSystem.getInfoAsync(FileSystem.documentDirectory + 'SQLite')).exists) {
    await FileSystem.makeDirectoryAsync(FileSystem.documentDirectory + 'SQLite');
  }
  await FileSystem.downloadAsync(
    Asset.fromModule(databaseFile).uri,
    FileSystem.documentDirectory + `SQLite/${databaseName}.db`,
  );
  return SQLite.openDatabase(`${databaseName}.db`);
};

export const conn = {
  file: (dbFile, dbName) => openDBfile(dbFile, dbName),
  init,
  tx,
};
