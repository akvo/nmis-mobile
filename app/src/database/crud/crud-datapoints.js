import { conn, query } from '../';

const db = conn.init;

const selectDataPointById = async ({ id }) => {
  const { rows } = await conn.tx(db, query.read('datapoints', { id }), [id]);
  if (!rows.length) {
    return {};
  }
  const current = rows._array[0];
  return {
    ...current,
    json: JSON.parse(current.json.replace(/''/g, "'")),
  };
};

const dataPointsQuery = () => {
  return {
    selectDataPointById,
    selectDataPointsByFormAndSubmitted: async ({ form, submitted }) => {
      const { rows } = await conn.tx(db, query.read('datapoints', { form, submitted }), [
        form,
        submitted,
      ]);
      if (!rows.length) {
        return [];
      }
      return rows._array;
    },
    selectSubmissionToSync: async () => {
      const submitted = 1;
      const syncedAt = null;
      const { rows } = await conn.tx(db, query.read('datapoints', { submitted, syncedAt }), [
        submitted,
        syncedAt,
      ]);
      if (!rows.length) {
        return [];
      }
      return rows._array;
    },
    saveDataPoint: async ({ form, user, name, submitted, duration, json }) => {
      const insertQuery = query.insert('datapoints', {
        form,
        user,
        name,
        submitted,
        duration,
        createdAt: new Date().toISOString(),
        submittedAt: submitted ? new Date().toISOString() : null,
        json: json ? JSON.stringify(json).replace(/'/g, "''") : null,
      });
      return await conn.tx(db, insertQuery, []);
    },
    updateDataPoint: async ({ id, name, submitted, duration, submittedAt, syncedAt, json }) => {
      const updateQuery = query.update(
        'datapoints',
        { id },
        {
          name,
          submitted,
          duration,
          submittedAt,
          syncedAt,
          json: json ? JSON.stringify(json).replace(/'/g, "''") : null,
        },
      );
      return await conn.tx(db, updateQuery, [id]);
    },
  };
};

const crudDataPoints = dataPointsQuery();

export default crudDataPoints;
