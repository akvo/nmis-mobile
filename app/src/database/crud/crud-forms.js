import { conn, query } from '../';

const db = conn.init;

const formsQuery = () => {
  return {
    selectLatestFormVersion: async () => {
      const latest = 1;
      const { rows } = await conn.tx(db, query.read('forms', { latest }), [latest]);
      if (!rows.length) {
        return [];
      }
      return rows._array;
    },
    selectFormByIdAndVersion: async ({ id: formId, version }) => {
      const { rows } = await conn.tx(db, query.read('forms', { formId, version }), [
        formId,
        version,
      ]);
      if (!rows.length) {
        return false;
      }
      return rows._array[0];
    },
    addForm: async ({ id: formId, version, formJSON }) => {
      const insertQuery = query.insert('forms', {
        formId: formId,
        version: version,
        latest: 1,
        name: formJSON?.name || null,
        json: formJSON ? formJSON : null,
        createdAt: new Date().toISOString(),
      });
      return await conn.tx(db, insertQuery, []);
    },
    updateForm: async ({ id: formId, latest = 0 }) => {
      // update latest to false
      const updateQuery = query.update('forms', { formId }, { latest: latest });
      return await conn.tx(db, updateQuery, [formId]);
    },
  };
};

const crudForms = formsQuery();

export default crudForms;
