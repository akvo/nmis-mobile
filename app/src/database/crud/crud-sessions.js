import { conn, query } from '../';

const db = conn.init;

const sessionsQuery = () => {
  return {
    selectLastSession: async () => {
      const { rows } = await conn.tx(db, query.read('sessions', []));
      if (!rows.length) {
        return false;
      }
      return rows._array[rows.length - 1];
    },
    addSession: async (data = { token: '' }) => {
      return await conn.tx(db, query.insert('sessions', data));
    },
  };
};

const crudSessions = sessionsQuery();

export default crudSessions;
