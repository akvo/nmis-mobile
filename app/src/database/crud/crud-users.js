import { conn, query } from '../';

const db = conn.init;

const usersQuery = () => {
  return {
    getActiveUser: async () => {
      try {
        const active = 1;
        const { rows } = await conn.tx(db, query.read('users', { active }), [active]);
        if (!rows.length) {
          return false;
        }
        return rows._array[0];
      } catch (error) {
        console.error('Get users', error);
        return false;
      }
    },
  };
};

const crudUsers = usersQuery();

export default crudUsers;
