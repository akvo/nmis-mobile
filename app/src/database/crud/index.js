export { default as crudSessions } from './crud-sessions';
export { default as crudForms } from './crud-forms';
export { default as crudUsers } from './crud-users';
