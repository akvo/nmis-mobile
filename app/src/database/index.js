export * from './conn';
export * from './query';
