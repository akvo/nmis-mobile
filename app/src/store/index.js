export * from './app';
export * from './auth';
export * from './buildParams';
export * from './forms';
export * from './users';
