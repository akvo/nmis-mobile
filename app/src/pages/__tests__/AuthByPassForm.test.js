import React from 'react';
import renderer from 'react-test-renderer';
import AuthByPassFormPage from '../AuthByPassForm';
import api from '../../lib/api';
import { UIState, UserState } from '../../store';
import { render, fireEvent, act, renderHook, waitFor } from '@testing-library/react-native';
import { Platform } from 'react-native';
import { useNavigation } from '@react-navigation/native';
import { crudUsers } from '../../database/crud';

jest.mock('../../lib/api');
jest.mock('../../database/crud');

describe('AuthByPassForm', () => {
  test('it renders correctly', () => {
    const { result: navigationRef } = renderHook(() => useNavigation());
    const navigation = navigationRef.current;
    const tree = renderer.create(<AuthByPassFormPage navigation={navigation} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('it should not download forms when offline', async () => {
    const { result: navigationRef } = renderHook(() => useNavigation());
    const navigation = navigationRef.current;
    api.post.mockImplementation(() => Promise.resolve({ data: { formsUrl: [] } }));

    render(<AuthByPassFormPage navigation={navigation} />);

    act(() => {
      UIState.update((s) => {
        s.online = false;
      });
    });
    expect(api.get).not.toHaveBeenCalled();
  });

  it('it should download forms when online', async () => {
    const { result: navigationRef } = renderHook(() => useNavigation());
    const navigation = navigationRef.current;
    api.get.mockImplementation(() => Promise.resolve({ data: { formsUrl: [] } }));

    render(<AuthByPassFormPage navigation={navigation} />);

    act(() => {
      UIState.update((s) => {
        s.online = true;
      });
    });

    expect(api.get).toHaveBeenCalledWith('/forms');
  });

  it('it should navigate to add user if no user defined after form downloaded', async () => {
    const { result: navigationRef } = renderHook(() => useNavigation());
    const navigation = navigationRef.current;
    const mockListForms = {
      message: 'Success',
      formsUrl: [],
    };

    api.getConfig.mockImplementation(() => ({ baseURL: 'http://example.com' }));
    api.get.mockImplementation(() => Promise.resolve({ data: mockListForms }));

    render(<AuthByPassFormPage navigation={navigation} />);

    act(() => {
      UIState.update((s) => {
        s.online = true;
      });
    });

    expect(api.get).toHaveBeenCalledWith('/forms');
    await waitFor(() => expect(navigation.navigate).toHaveBeenCalledWith('AddUser'));
  });

  it('it should navigate to home if user defined after form downloaded', async () => {
    const { result: navigationRef } = renderHook(() => useNavigation());
    const navigation = navigationRef.current;
    const mockGetData = {
      message: 'Success',
      formsUrl: [
        {
          id: 1,
          url: '/forms/1',
          version: '1.0.0',
        },
      ],
    };
    const mockUser = { id: 1, name: 'John Doe', password: 'qwerty' };
    api.get.mockImplementation(() => Promise.resolve({ data: mockGetData }));
    crudUsers.getActiveUser.mockImplementation(() => Promise.resolve(mockUser));
    const { result: userStateRef } = renderHook(() => UserState.useState((s) => s));

    render(<AuthByPassFormPage navigation={navigation} />);

    act(() => {
      UIState.update((s) => {
        s.online = true;
      });
    });

    expect(api.get).toHaveBeenCalledWith('/forms');
    await waitFor(() => {
      const {
        id: userIdState,
        name: userNameState,
        password: userPasswordState,
      } = userStateRef.current;
      expect(userIdState).toEqual(mockUser.id);
      expect(userNameState).toEqual(mockUser.name);
      expect(userPasswordState).toEqual(mockUser.password);
      expect(navigation.navigate).toHaveBeenCalledWith('Home');
    });
  });

  it('it should be error', async () => {
    const { result: navigationRef } = renderHook(() => useNavigation());
    const navigation = navigationRef.current;
    const mockErrorData = { message: 'Failed' };
    api.get.mockImplementation(() =>
      Promise.reject({ response: { ...mockErrorData, status: 400 } }),
    );
    crudUsers.getActiveUser.mockImplementation(() => Promise.resolve(mockUser));
    const { result: userStateRef } = renderHook(() => UserState.useState((s) => s));

    render(<AuthByPassFormPage navigation={navigation} />);

    act(() => {
      UIState.update((s) => {
        s.online = true;
      });
    });

    expect(api.get).toHaveBeenCalledWith('/forms');
  });
});
