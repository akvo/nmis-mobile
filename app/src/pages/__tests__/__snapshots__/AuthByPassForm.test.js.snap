// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthByPassForm it renders correctly 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#f9fafb",
      "display": "flex",
      "flex": 1,
      "flexDirection": "column",
      "gap": 36,
      "justifyContent": "center",
      "paddingHorizontal": 16,
    }
  }
  testID="center-layout"
>
  <View
    accessibilityIgnoresInvertColors={true}
    style={
      {
        "backgroundColor": "#e5e7eb",
        "borderRadius": 4,
        "height": 110,
        "overflow": "hidden",
        "position": "relative",
        "width": 110,
      }
    }
  >
    <Image
      onLoad={[Function]}
      source={
        {
          "uri": "uri",
        }
      }
      style={
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
      testID="image-component"
      theme={
        {
          "colors": {
            "background": "#ffffff",
            "black": "#242424",
            "disabled": "hsl(208, 8%, 90%)",
            "divider": "#bcbbc1",
            "error": "#ff190c",
            "grey0": "#393e42",
            "grey1": "#43484d",
            "grey2": "#5e6977",
            "grey3": "#86939e",
            "grey4": "#bdc6cf",
            "grey5": "#e1e8ee",
            "greyOutline": "#bbb",
            "platform": {
              "android": {
                "error": "#f44336",
                "grey": "rgba(0, 0, 0, 0.54)",
                "primary": "#2196f3",
                "searchBg": "#dcdce1",
                "secondary": "#9C27B0",
                "success": "#4caf50",
                "warning": "#ffeb3b",
              },
              "default": {
                "error": "#ff3b30",
                "grey": "#7d7d7d",
                "primary": "#007aff",
                "searchBg": "#dcdce1",
                "secondary": "#5856d6",
                "success": "#4cd964",
                "warning": "#ffcc00",
              },
              "ios": {
                "error": "#ff3b30",
                "grey": "#7d7d7d",
                "primary": "#007aff",
                "searchBg": "#dcdce1",
                "secondary": "#5856d6",
                "success": "#4cd964",
                "warning": "#ffcc00",
              },
              "web": {
                "error": "#ff190c",
                "grey": "#393e42",
                "primary": "#2089dc",
                "searchBg": "#303337",
                "secondary": "#ca71eb",
                "success": "#52c41a",
                "warning": "#faad14",
              },
            },
            "primary": "#2089dc",
            "searchBg": "#303337",
            "secondary": "#ad1457",
            "success": "#52c41a",
            "warning": "#faad14",
            "white": "#ffffff",
          },
          "spacing": {
            "lg": 12,
            "md": 8,
            "sm": 4,
            "xl": 24,
            "xs": 2,
          },
        }
      }
      transitionDuration={360}
    />
    <View
      accessibilityElementsHidden={true}
      collapsable={false}
      importantForAccessibility="no-hide-descendants"
      pointerEvents="none"
      style={
        {
          "bottom": 0,
          "left": 0,
          "opacity": 1,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#bdbdbd",
            "justifyContent": "center",
          }
        }
        testID="RNE__Image__placeholder"
      />
    </View>
    <View
      style={{}}
      testID="RNE__Image__children__container"
    />
  </View>
  <View>
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "center",
        }
      }
    >
      <ActivityIndicator
        color="#2089dc"
        size="large"
        style={
          {
            "marginVertical": 20,
          }
        }
        testID="Dialog__Loading"
      />
    </View>
    <Text
      accessibilityRole="text"
      style={
        {
          "color": "#242424",
          "fontStyle": "italic",
          "textAlign": "center",
        }
      }
    >
      Fetching data
    </Text>
  </View>
</View>
`;
